src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,532 @@
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected static void WriteVersion(JsonWriter writer, string key, Version version)
+        {
+            writer.WriteKey(key);
+            writer.WriteObjectStart();
+            writer.Write("major", version.Major);
+            writer.Write("minor", version.Minor);
+            writer.Write("build", version.Build);
+            writer.Write("revision", version.Revision);
+            writer.WriteObjectEnd();
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
+        {
+            writer.WriteKey(key);
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
+        {
+            if (type is null)
+            {
+                writer.WriteNull();
+                return;
+            }
+
+            writer.WriteObjectStart();
+            writer.Write("fullName", type.FullName);
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.Write("mvid", GetGuidValue(type.Assembly.ManifestModule.ModuleVersionId));
+            writer.WriteObjectEnd();
+        }
+
+        private (JsonWriter, PooledStringBuilder) CreateWriter()
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            var writer = new StringWriter(builder);
+            return (new JsonWriter(writer), builder);
+        }
+
+        internal string GetKey(
+            CompilationOptions compilationOptions,
+            ImmutableArray<SyntaxTreeKey> syntaxTrees,
+            ImmutableArray<MetadataReference> references,
+            ImmutableArray<byte> publicKey,
+            ImmutableArray<AdditionalText> additionalTexts,
+            ImmutableArray<DiagnosticAnalyzer> analyzers,
+            ImmutableArray<ISourceGenerator> generators,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            EmitOptions? emitOptions,
+            DeterministicKeyOptions options,
+            CancellationToken cancellationToken)
+        {
+            additionalTexts = additionalTexts.NullToEmpty();
+            analyzers = analyzers.NullToEmpty();
+            generators = generators.NullToEmpty();
+
+            var (writer, builder) = CreateWriter();
+
+            writer.WriteObjectStart();
+
+            writer.WriteKey("compilation");
+            WriteCompilation(writer, compilationOptions, syntaxTrees, references, publicKey, pathMap, options, cancellationToken);
+            writer.WriteKey("additionalTexts");
+            writeAdditionalTexts();
+            writer.WriteKey("analyzers");
```

Comment:
Do we have tests with non-empty analyzers? 
Same question for generators.

Reply:
Fixed